### ISSUE TITLE
i18n / Интернационализация

### ISSUE THREAD

**Issue by aeifn** [PastVu/pastvu#257]
Представляется логичным перенести языковую составляющую в мастер ветку, тем самым избавиться от необходимости синхронизировать ветку `en` с мастером и собирать два разных образа приложения.

Зачатки такого подхода наблюдаются, например, здесь:
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/13631515/98915023-6b80b680-24da-11eb-9b6e-eead648dabf4.png">

И здесь:
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/13631515/98915077-7b989600-24da-11eb-817d-a20932adf39a.png">

**Comment by mkgrgis** [comment]
Что если хранить такие вещи в статической ветке Монги с разметкой согласно ISO 639 ? 
`about_title_text: { ..., eng:"About", rus: "О проекте" ... }`
Для начала хорошо бы вынести в JSON файл или файлы, где рядом вести ряды на доступных языках в одинаковом порядке по латинскому алфавиту языковых кодов.
Другой вариант - доверится средствам ОС и работать с gettext через [пакет](https://www.npmjs.com/package/gettext). Только преимуществ простоты редактирования в таком случае нет.

**Comment by paul-k-pastvu** [comment]
На самом деле, не нужно вообще так уж держаться за эти строки и пытаться что-то организовать у себя.
Гораздо лучше всё вынести в языковые .po файлы и поднять любую систему перевода. Или подключиться к https://translatewiki.net или https://www.transifex.com/ - что во всех смыслах было бы эффективнее.

**Comment by mkgrgis** [comment]
> лучше всё вынести в языковые .po файлы

Так это и есть gettext вариант в чистом виде, [см.](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html) В таком случае [пакета](https://www.npmjs.com/package/gettext)  не избежать.

**Comment by aeifn** [comment]
> > лучше всё вынести в языковые .po файлы
> 
> Так это и есть gettext вариант в чистом виде, [см.](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html) В таком случае [пакета](https://www.npmjs.com/package/gettext) не избежать.

Gettext представляется отличным выбором, и внедрить его технически совсем не трудно. Но потребуется кропотливая работа по аккуратной замене сообщений на соответствующие функции и подготовке языковых файлов.
Предлагаю подключить пакет и попробовать перевести отдельную небольшую часть приложения для пробы.

**Comment by aeifn** [comment]
Лучше использовать этот пакет: https://www.npmjs.com/package/node-gettext
Он более популярен и поддерживается

**Comment by anvaruk** [comment]
>  ... Но потребуется кропотливая работа по аккуратной замене сообщений на соответствующие функции и подготовке языковых файлов.

я могу заняться если инструкцию дадите

**Comment by mkgrgis** [comment]
@anvaruk сначала нужно попробовать просто добиться в изолированной node js работы [пакета](https://www.npmjs.com/package/node-gettext) с переводом и перевести 1-2 сообщения через создание файлов `.po` на 2-3 языка. Если получтся, то @aeifn должен определить какие вызовы с формированеим надписей подлежат первоочередной замене на вызовы `getText`. А там только `.po` дополняй и наращивай.

**Comment by aeifn** [comment]
Доступность функции gettext в шаблонах можно обеспечить отсюда:
https://github.com/PastVu/pastvu/blob/c47487f6e3d509bd1142bb61fca62e7afb199933/app.js#L92-L98

**Comment by kabalin** [comment]
> Лучше использовать этот пакет: https://www.npmjs.com/package/node-gettext
> Он более популярен и поддерживается

На что нужно обратить внимание:
* Как формировать JSON для использования на бэкэнде - компилировать отдельной задачей в grunt (и включать в сборку) или компилировать при каждом запуске приложения?
* Как передавать JSON c набором строк клиенту - через  вебсервис подгружать для выбранной локали или компилировать отдельной задачей в grunt и включать в сборку в поддиректорию в `public`? Вероятно при делать это сборке оптимальный вариант (как считает @aeifn), для разработчика альтернативный способ компиляции будет ручной запуск команды `grunt po2json`.
* Со стороны клиента у помимо js, еще шаблоны:
  * В шаблонах есть статический текст (типа [такого](https://github.com/PastVu/pastvu/blob/master/views/module/common/foot.pug#L4)), его нужно как-то преобразовывать.
  * В шаблонах есть knockout, где текст является частью data-bind (типа [такого](https://github.com/PastVu/pastvu/blob/master/views/module/photo/photo.pug#L89)), здесь нужен отдельный обработчик для knockout (этот для [примера](https://github.com/vicary/knockout-i18n), и [еще один](https://github.com/leMaik/i18next-ko) для i18next, там можно подчерпнуть идеи как встроить в knockout).

Чтобы с чего то начать, шаги могут быть следующими:
1. Взять какой-то шаблон
2. Создать po файл со всеми строками из шаблона (в `./locales` ?).
3. Прикрутить парсер к Gruntfile (см https://github.com/mikeedwards/po2json и https://www.npmjs.com/package/grunt-po2json)
4. Сгенерировать json с текстом из po файла.
5. В шаблоне получить текст из json используя [node-gettext](https://www.npmjs.com/package/node-gettext) и https://github.com/PastVu/pastvu/issues/257#issuecomment-740253183

**Comment by yuriy-zhilovets2** [comment]
Стандартный механизм использования gettext такой:

1) Все строки в коде заменяются вызовом функции _
_("Это строка")

2) Отдельный парсер, запускаемый при изменения в коде, находит эти строки и переносит в каталог строк (отдельный файл)

3) В специальном редакторе найденные строки переводятся на поддерживаемые языки и сохраняются в файле .po

4) Редактор компилирует этот файл в формат .mo, файл подкладывается в дерево

5) Функция _ во время работы читает строки из файла .mo

Тонкости:
1) Строки могут быть не только в коде .js, но и в других местах. Соответственно, нужен отдельный парсер и механизм вызова
2) Файлы .mo получаются довольно большого размера, поэтому имеет смысл разделить строки в серверном и в клиентском коде, и в браузер грузить только один файл для клиентских строк

Я могу взяться за перевод строк, но нужно определиться:

1. Какой пакет будет использоваться
2. Где еще могут находиться строки (например, в сообщениях об ошибках)

Последовательность действий:
1. Начать со строк в коде, отметить их через _
2. Собрать каталог, проверить, что строки транслируются правильно
3. Добавить английский перевод и код переключения на него
4. Разобраться с остальными строками

**Comment by aeifn** [comment]
> * Строки могут быть не только в коде .js, но и в других местах. Соответственно, нужен отдельный парсер и механизм вызова

По идее, все строки должны быть в js-коде.

> * Файлы .mo получаются довольно большого размера, поэтому имеет смысл разделить строки в серверном и в клиентском коде, и в браузер грузить только один файл для клиентских строк

Возможно, лучше просто собирать несколько версий фронта на разных языках (строки будут подставляться на этапе сборки).

**Comment by aeifn** [comment]
Предлагаю использовать этот фреймворк: https://github.com/i18next/i18next

**Comment by kabalin** [comment]
> Предлагаю использовать этот фреймворк: https://github.com/i18next/i18next

Идея хорошая, но как я понимаю он заточен под runtime, у нас строки еще и на стороне сервера есть и в клиенте в нескольких вариантах (js, шаблоны). Шаблон компилируется при сборке, там динамически строки менять не получится (если только их не будет подставлять JS библиотека при показе), в общем есть вопросы. Мне кажется встраивание языка на этапе компиляции и отдельные инстансы приложения для разных языков - более понятный вариант.

Отдельно скажу, что я так же не уверен насчет po/mo файлов и gettext. Хороший пример реализации перевода в Docusaurus, там в качестве фреймворка используется [formatjs](https://formatjs.io/), в качестве формата файлов перевода [Chrome.i18n](https://docusaurus.io/docs/i18n/introduction#json-files) который имеет дополнительное поле description (помогает переводчику) и [интеграцию с сервисом Crowdin](https://docusaurus.io/docs/i18n/crowdin) обеспечивающим интерфейс для переводчика (бесплатно для open source проектов). Мне кажется нужно что-то похожее делать с удобным форматированием и на основе стандартов (у i18next как я понял свой стандарт форматирования для множественного числа, даты, вставки переменных и т.д., а formatjs использует [ICU Message syntax](https://unicode-org.github.io/icu/userguide/format_parse/messages) поддерживаемый и другими фрейвоками, i18n-node в частности).

**Comment by klimashkin** [comment]
Согласен, использую formatjs c ICU/CLDR c 2014 года на других проектах.
Это Unocode стандарт для формата строк

**Comment by paul-k-pastvu** [comment]
Это, наверное, тоже тема для раздела "дискуссии"?

**Comment by aeifn** [comment]
> Это, наверное, тоже тема для раздела "дискуссии"?

То, что делать это нужно, спора нет. Дискуссия тут может касаться лишь технических моментов реализации.

**Comment by aeifn** [comment]
Closed by https://github.com/PastVu/pastvu/pull/764
